### ISSUE TITLE
Get told that Roslyn is going to remove references when running a Code Fix

### ISSUE THREAD

**Issue by davkean** [dotnet/roslyn#16083]
This is a timing thing that I've seen a few times, if I run a code fix *before* the solution has finished loading (I believe that's related), I see the following:

![image](https://cloud.githubusercontent.com/assets/1103906/21446113/23168a4c-c916-11e6-8eaa-572f3ddde76e.png)

There's not a 100% repro here, so play around with the ProjectSystem solution.

**Comment by davkean** [comment]
tag @CyrusNajmabadi

**Comment by davkean** [comment]
What's even weirder is that this project doesn't even reference those projects.

**Comment by CyrusNajmabadi** [comment]
@pilchie, i've been able to repro this while the solution is loading.  What is the right way to know if the solution is loading?  We likely want to just suppress lightbulbs entirely while that is happening.

**Comment by Pilchie** [comment]
http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/VisualStudioProjectTracker.cs,5fd643f3ca41cd15

**Comment by CyrusNajmabadi** [comment]
@Pilchie Does that work for DPL/AnyCode?  Thanks!

**Comment by Pilchie** [comment]
Nothing works for anycode, but we don't offer lightbulbs there anyway.

For DPL we don't have a good answer yet, but it's something we need to figure out soon.

**Comment by CyrusNajmabadi** [comment]
I'm not sure what to do here.   If we can provide some sort of service that can effectively answer this question, then that would be great.  But i don't know how to make that service myself.

**Comment by heejaechang** [comment]
interesting. so is it because project is fully loaded but solution is not (I am including design time build not finished as part of solution not fully loaded yet)

anyway, there could be many fixes that can work even if solution is not fully loaded I think that was why we let third party analyzers to work even if project is in broken state.

shouldn't we follow same principle here? and rather than engine do something, each fix decide whether they will provide fix or not? like what we decided for diagnostics.

if we let diagnostic to be reported from third party and then turn off fix, that would moot the reason we enabled third party analyzers.

I think we should provide a way for users to check state of various error cases such as workspace.solutionfullyloaded, or project.fullyloaded. and let fix to decide whether they will offer fix or not.

**Comment by CyrusNajmabadi** [comment]
> anyway, there could be many fixes that can work even if solution is not fully loaded I think that was why we let third party analyzers to work even if project is in broken state.

No fixes can work.  At least none that are DocumentChange/ProjectChange/SolutionChange actions.  That's because their resultant SolutionSnapshots will always break the final solution by undoing changes.

Also, we can't even apply those fixes as TryApplyChanges will immediately bail as the snapshots one is trying to apply isn't branched off the .CurrentSolution for the workspace.

**Comment by heejaechang** [comment]
hmm.. then what is the point of us enabling third party analyzers on design time build not done yet case?

**Comment by CyrusNajmabadi** [comment]
> hmm.. then what is the point of us enabling third party analyzers on design time build not done yet case?

That's exactly the bug that this issue is reporting :)

**Comment by heejaechang** [comment]
well, we used to not report on such case (before 15.3), and we explicitly made it possible in 15.3 and said, responsibility is on author not the engine. 

I think we should follow same principle on both fix and analyzer whatever way we decide.

...

about solution keep changing issue. i dont think that is issue due to solution loading, it certainly can happen more frequently, but same thing can happen in completely different situation such as autogenerated files being updated, source control synching new files and etc.

that is just our underlying issue of how we implemented Applying and Diffing.

**Comment by sharwell** [comment]
> hmm.. then what is the point of us enabling third party analyzers on design time build not done yet case?

The bug is that we fail to display and apply the code fixes that are provided. The fact that we display code fixes that are available is not a bug.

**Comment by heejaechang** [comment]
but that is not what Cyrus is saying. in current system, it will always fail.

**Comment by sharwell** [comment]
@heejaechang I edited my post. The current system is not yet able to handle this case, but we need to change that.

**Comment by heejaechang** [comment]
sure. I am just saying we should use same principle on both diagnostics and fixes. either engine takes care of it or let author to take care of it in error situation.

one let author be in charge and the other engine in charge seems breaking each other.

**Comment by jasonmalinowski** [comment]
Chiming in here since it got discussed internally: ignoring for a moment the entire question of _when_ to even start computing fixes, my guess is the bad diffs appear because we compute the final solution, and then when we show the diff we are diffing not against the solution we originally gave to the code fixer but rather Workspace.CurrentSolution. This is a bug that we should fix no matter what -- as the project system increasingly goes asynchronous we need to be resilient to things changing.

**Comment by heejaechang** [comment]
but we also need to fix that TryApplyChange accepting changes around CurrentSolution as well. otherwise, diff (preview) they see and result they get might be different.

**Comment by CyrusNajmabadi** [comment]
> but we also need to fix that TryApplyChange accepting changes around CurrentSolution as well. otherwise, diff (preview) they see and result they get might be different.

Note: i believe that we actually bail out in this case.  However, this is also a bad experience as you can invoke a lightbulb and then have nothing happen. This happens to me a *lot* when i want to open VS to fix up git-merge-conflicts.  I get to the file then tyr to use hte lightbulb, and nothing actually changes.  

If we can't apply a change when you invoke things, we should at least give a good error message like:

"Could not apply light-bulb change.  The solution changed in between computing the fix and applying it.  Application was canceled to prevent unwanted changes from happening."

**Comment by heejaechang** [comment]
yep. we know it checks given solution is a fork of current solution and only accept those. what I meant was we need to make it to accept solution that is not fork of current solution. and also, it shouldn't just diff from current solution as well.

it should be diff between forked solution (code fix) and base solution that is applied to current solution.

easy to say, hard to actually do fully. but I think most of common cases such as (changing text on a file especially open file), should be quite safe to do.

**Comment by CyrusNajmabadi** [comment]
I could be convinced that this would ok *solely* for the case of a single file being edited :)

**Comment by heejaechang** [comment]
I wasn't suggesting solution. I was saying it would be nice, or time to make TryApplyChanges to be a bit more smarter. since now we are going to go really async world where world can always change in the background for whatever reason.

or we just bring back synchronous world we used to implicitly do through UI thread. 
for example, (again not suggesting solution) explicitly create a queue where workspace change can happen, and all workspace change go trough this queue like UI thread, but not implicit queue such as windows message queue (which is what we actually mean by UI thread), but have explicit WorkspaceChangeQueue. and features such as code fix can block, that queue (which we did implicitly by blocking UI message queue) when Invoke is called so that workspace can't mutate while it is calculating fix.
